plugins/rs_lisp.py

```diff
@@ -0,0 +1,472 @@
+            t = read_from_tokens(tokens)
+            if t != '':
+                l.append(t)
+        tokens.pop(0)
+        return l
+
+    elif ';' == token:
+        l = []
+        l.append(token)
+        while tokens[0] != '\n':
+            l.append(read_from_tokens(tokens))
+        new_line = tokens.pop(0)
+        l.append(new_line)
+        string = l_restore(" ".join(l))
+        return ['quote', string]
+
+    elif ')' == token:
+        raise CustomCommandSyntaxError('unexpected )')
+    else:
+        return atom(re.sub(r"^\s+|\s+$", "", token))
+
+
+def atom(token: str):
+    try:
+        return int(token)
+    except ValueError:
+        try:
+            return float(token)
+        except ValueError:
+            try:
+                return is_positive(token)
+            except CommandSyntaxError:
+                return token.lower()
+
+
+# =============================================================================================================
+# Evaluator
+
+# A user-defined Scheme procedure.
+class Procedure(object):
+    def __init__(self, parms, body, env):
+        self.parms, self.body, self.env = parms, body, env
+
+    def __call__(self, *args):
+        return lisp_eval(self.body, Env(self.parms, args, self.env))
+
+
+class Env(dict):
+    def __init__(self, parms=(), args=(), outer=None, max_runtime=0):
+        self.update(zip(parms, args))
+        self.outer = outer
+        self.timestamp = time()
+        self.max_runtime = max_runtime
+
+    # Find the innermost Env where var appears.
+    def find(self, var):
+        if var in self:
+            return self
+        elif self.outer:
+            return self.outer.find(var)
+        else:
+            raise CustomCommandSyntaxError(f'undefined var {var}')
+
+
+def get_args(args: list) -> (list, dict):
+    t_list = [*args]
+    t_dict = OrderedDict()
+    for n, a in [*enumerate(t_list)][::-1]:
+        if type(a) == str and a.startswith(':'):
+            try:
+                t_dict[a[1:]] = t_list.pop(n + 1)
+            except IndexError:
+                raise CustomCommandSyntaxError(f"supplied argument {a} given without value")
+            del t_list[n]
+    return t_list, OrderedDict(reversed(t_dict.items()))
+
+
+def _str(*args):
+    if not args:
+        return
+    elif len(args) == 1:
+        return str(args[0])
+    else:
+        try:
+            return getattr(str, args[0])(*args[1:])
+        except AttributeError:
+            raise CustomCommandSyntaxError(f'str does not have method {args[1]}')
+
+
+def eztime(*args):
+    time = datetime.datetime.utcnow()
+
+    if args:
+        strf = args[0] if args[0] else "%Y-%m-%d @ %H:%M:%S"
+        if len(args) > 1:
+            o_time = re.match(r"(?P<h>-?\d*):(?P<m>-?\d*):(?P<s>-?\d*)", args[1])
+            if o_time:
+                o_time = o_time.groupdict()
+                delta = datetime.timedelta(hours=int(o_time['h']) if o_time['h'] else 0,
```

Comment:
This will raise a `KeyError`, use `.get` instead

Reply:
It will not raise a KeyError because there is *always* `o_time['h']`. It's just an empty string some times since that fulfills the `-?\d*` pattern.